File: src/algorithms.py
"""
Optimized code for calculating the likelihood of an exclusive relationship.
Read the whole project at https://tarro.work/code/the-hannah-montana-problem
"""
import math
from typing import List


def likelihood_of_an_exclusive_relationship_not_optimized(A: List[int], N: int) -> float:
    """The probability that no samples of sizes `A` share a value from a population of size `N`.

    Is not optimized in any way.
    """
    p: float = 1.0
    for i, a in enumerate(A):
        p *= math.comb(N - sum(A[:i]), a) / math.comb(N, A[i])
    return 1.0 - p


def likelihood_of_an_exclusive_relationship_algebraically_optimized(A: List[int], N: int) -> float:
    """The probability that no samples of sizes `A` share a value from a population of size `N`.
    
    Uses an optimized formula.
    """
    _sum: int = 0
    numer: int = 1
    denom: int = 1
    for a in A:
        # This multiplication is comparable to the H function
        denom *= math.prod(range(N - a + 1, N + 1))
        _sum += a

    # This multiplication is comparable to the H function
    numer = math.prod(range(N - _sum + 1, N + 1))
    return 1.0 - numer / denom


def likelihood_of_an_exclusive_relationship_completely_optimized(A: List[int], N: int) -> float:
    """The probability that no samples of sizes `A` share a value from a population of size `N`.

    Uses an optimized formula and algorithm.
    """
    A_sorted: List[int] = sorted(A)
    _sum: int
    numer: int
    denom: int

    # Denominator Computations
    h: int = 1
    last_h = None
    initial_x_of_last_h = N - A_sorted[0] + 1

    # calculating h for the smallest sample size in the sequence
    h = math.prod(range(initial_x_of_last_h, N + 1))

    _sum = A_sorted[0]
    denom = h
    last_h = h

    for a in A_sorted[1:]:
        h = last_h
        initial_x = N - a + 1
        # Calculating h up to the last x used in an h calculation
        h *= math.prod(range(initial_x, initial_x_of_last_h))
        denom *= h
        _sum += a

        # Update memory
        last_h = h
        initial_x_of_last_h = initial_x

    # Numerator Computations
    numer = last_h
    # Calculating h up to the last x used in an h calculation
    numer *= math.prod(range(N - _sum + 1, initial_x_of_last_h))

    return 1.0 - numer / denom

File: algorithms/algorithms.rs
/**
 * Optimized code for calculating the likelihood of an exclusive relationship.
 * Read the whole project at https://tarro.work/code/the-hannah-montana-problem
 */

/// Computes a factorial.
/// 
/// # Arguments
/// 
/// * `n` - A positive integer to calculate a factorial for
/// 
fn factorial(n: &u64) -> u64 {
    (1..=*n).product::<u64>()
}

/// Counts the number of ways of choosing `r` unordered outcomes from `n` possibilities.
/// 
/// # Arguments
/// 
/// * `n` - The population size
/// * `r` - The sample size
/// 
fn comb(n: &u64, r: &u64) -> u64 {
    (n - r + 1..=*n).product::<u64>() / factorial(r)
}

/// The probability that no samples of sizes `samples` share a value from a population of size `population`.
/// 
/// Is not optimized in any way.
/// 
/// # Arguments
/// 
/// * `samples`     - A vector of sample sizes
/// * `population`  - The population size
/// 
pub fn likelihood_of_an_exclusive_relationship_not_optimized(samples: &Vec<u64>, population: &u64) -> f64 {
    let mut p: f64 = 1.0;
    let mut sum: u64 = 0;
    for a in samples {
        p *= comb(&(population - sum), &a) as f64 / comb(population, a) as f64;
        sum += a;
    }
    1.0 - p
}

/// The probability that no samples of sizes `samples` share a value from a population of size `population`.
/// 
/// Uses an optimized formula.
/// 
/// # Arguments
/// 
/// * `samples`     - A vector of sample sizes
/// * `population`  - The population size
/// 
pub fn likelihood_of_an_exclusive_relationship_algebraically_optimized(samples: &Vec<u64>, population: &u64) -> f64 {
    let mut sum: u64 = 0;
    let mut denom: u64 = 1;
    let numer: u64;
    for a in samples {
        denom *= ((population - a + 1)..=*population).product::<u64>();
        sum += a;
    }
    numer = ((population - sum + 1)..=*population).product::<u64>();
    1.0 - (numer as f64 / denom as f64)
}

/// The probability that no samples of sizes `samples` share a value from a population of size `population`.
/// 
/// Uses an optimized formula and algorithm.
/// 
/// # Arguments
/// 
/// * `samples`     - A vector of sample sizes
/// * `population`  - The population size
/// 
pub fn likelihood_of_an_exclusive_relationship_completely_optimized(samples: &Vec<u64>, population: &u64) -> f64 {
    let mut sorted_samples: Vec<u64> = samples.to_vec();
    sorted_samples.sort();
    let mut sum: u64 = 0;

    let numer: u64;
    let mut denom: u64 = 1;

    let mut h: u64 = 1;
    let mut initial_x: u64;
    let mut initial_x_of_last_h: u64 = population + 1;

    for a in samples {
        initial_x = population - a + 1;
        h *= (initial_x..=initial_x_of_last_h - 1).product::<u64>();
        denom *= h;
        sum += a;

        initial_x_of_last_h = initial_x;
    }

    numer = h * (population - sum + 1..=initial_x_of_last_h - 1).product::<u64>();

    1.0 - (numer as f64 / denom as f64)
}

fn main() {
    let samples: Vec<u64> = vec![300, 2];
    let population: u64 = 1000;
    let result_no = likelihood_of_an_exclusive_relationship_not_optimized(&samples, &population);
    println!("no: {result_no}");
    let result_a = likelihood_of_an_exclusive_relationship_algebraically_optimized(&samples, &population);
    println!("a: {result_a}");
    let result_op = likelihood_of_an_exclusive_relationship_completely_optimized(&samples, &population);
    println!("op: {result_op}");
}